Swerve

adamv17 opened the pull request:
This pull request was created for the Galaxia Swerve.

The swerve has the following motors for each module (wheel):
-  775 redline, TalonSRX controller
-  TalonFX, built in controller

The TalonSRX moves the wheel to a target angle, the TalonFX spins the wheel.

The subsystem Swerve Drive receives in the constructor whether it is field oriented or not. The main method in the subsystem is holonomicDrive which receives the joystick inputs calculates the Vx and Vy wheel velocities, converts them to a polar coordinate system (V, θ).  For field oriented a rotation matrix is used. The size of the velocity and angle are than assigned to the corresponding wheels (SwerveModule).

The Commands:
- HolonomicDrive, default command, operates based on the three joystick values
- DriveForward, basic test of driving forward
- TurnInPlace, basic test of turning in place
- ResetPositions, resets encoder positions

Commit:
Add constant kF.

natymeitav committed:
Add Funnel's Ports.

natymeitav committed:
Add Funnel Constructor.

natymeitav committed:
Add setVelocity function to Funnel subsystem.

natymeitav committed:
Add Funnel's constants.

natymeitav committed:
Create StartFunnel Command.

natymeitav committed:
Make Funnel extend SubsystemBase.

natymeitav committed:
Add Funnel's commands to robotContainer.

natymeitav committed:
Document StartFunnel.

natymeitav committed:
document Funner.

natymeitav committed:
Format files.

orel-david committed:
Fix getTargetAngle method and utilize in code for smooth driving

orel-david committed:
Fix lock angle method

orel-david committed:
Add drive function

barel13 committed:
Add state space control dependency

It's also update the version of wpilib

barel13 committed:
Fix the configuration in the constructor

Both modules should have the same PID, because if the won't, most likely they won't move at the same rate

barel13 committed:
Create set, get and estimate velocity methods

The estimate velocity function give and approximation of the velocity based on the distance of the robot from the target, we need to get the function via an experiment and excel

barel13 committed:
Create set power and stop functions

the set power function used in the code in conjunction with the state space control

barel13 committed:
Create is ready function

the isReady function checks whether the flywheel is "close" enough to the desired velocity and whether the wheel is actually moves (MINIMAL_VELOCITY)

barel13 committed:
Create Flywheel subsystem

this class is responsible to encapsulate the wheels of the flywheel in one class and to calculate the power to shoot using state-space control

barel13 committed:
Create Ports and Flywheel nested class

barel13 committed:
Add flywheel modules and bind them into the flywheel subsystem

barel13 committed:
Add state-space to the code

I added the option to disable state-space in case it won't work, so we still have stable code for the season (always be prepared😂). Also, we never used state-space in the flywheel, so until it stable we need to be sure. We Use kalman and LQR because wpilib force us to use them, and most likely it won't harm

barel13 committed:
Create setPower function

barel13 committed:
Create setVelocity and stop functions in Flywheel subsystem

The set velocity function has 2 versions, one for state-space and the other is the "native" way (just PID). At the beginning of the state-space part we tell the state-space to update based on the velocity, afterwards we correct the model based on the velocity, and predict the next velocity we should give to the motor. Finally we power the motors

Commit:
Fix TOLERANCE constant.

Commit:
Change naming scheme.

Se-gol committed:
Update src/main/java/frc/robot/subsystems/climber/Climber.java

Co-authored-by: Dan Katzuv <31829093+katzuv@users.noreply.github.com>

Commit:
Add CRUISE_VELOCITY & ACCELERATION constants.

Commit:
Change name to ShiftGear.

Commit:
Change command to manage climb.

Commit:
Fix auto climb command.

Commit:
Fix docu and configs.

Commit:
Remove unnecessary line.

Commit:
Merge remote-tracking branch 'origin/climber' into climber

barel13 committed:
Create Shoot command

The shoot command is responsible for shooting the balls ti the desired location based on the velocity parameter. The velocity is determined via the "estimateVelocity" function, that by the time we fill it.

barel13 committed:
Implement Shoot command

barel13 committed:
Move estimateVelocity function from flywheel module to flywheel

Both motors should have the same velocities, so I think that this function is more related to the flywheel class, the class that contains all of the wheels

Commit:
Fix colour_wheel.

Commit:
Fix climber.

Commit:
Merge remote-tracking branch 'origin/climber' into climber

barel13 committed:
Add documentation to FlywheelModule

barel13 committed:
Fix documentation

p.s it's Saar

adamv17 committed:
Merge branch 'test-run' of https://github.com/Adam14boat/Swerve into swerve

barel13 committed:
Rename Flywheel class to Shooter and flywheel package to shooter

barel13 committed:
Add follow function to FlywheelModule

barel13 committed:
Remove State-Space from FlywheelModule class

the state-space is already in the Shooter class

barel13 committed:
Remove useStateSpace boolean from Shooter class and split the modules

We have already tested the state-space and is better than PID, so we chose to use it as the default in the shooter. Also, the modules are specific to our robot, so I chose to create an object for each module instead of using array. More then that, most likely that no one will create a shooter with more than 2 wheels in FRC.

barel13 committed:
Rename getAvgVelocity to getVelocity

The aux is following the main motor, so his velocity should be the same as the velocity of the main motor

barel13 committed:
Remove FlywheelModule class

I saw that most of the functions in Shooter class are copies of the functions in FlywheelModule class, so I chose to migrate the functions from the FlywheelModule class to Shooter.

barel13 committed:
Add VOLTAGE and ARBITRARY_FEED_FORWARD constants

barel13 committed:
Update Shooter class

Change the Port names to main and aux inverted, and removed the sensor use in the aux motor. Also, I add documentation to the Shooter class, and migrate 
from FlywheelModule to Shooter.

katzuv commented on src/main/java/frc/robot/Utils.java:
```diff
@@ -0,0 +1,67 @@
+     */
+    public static double[] cartesianToPolar(double x, double y) {
+        double[] polar = new double[2];
+        polar[0] = Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2));
```
Use [`Math.hypot`](https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#hypot-double-double-) instead.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,264 @@
+
+public class SwerveDrive extends SubsystemBase {
+
+    static double[][] M = new double[8][3];
```
Be more specific with the variable name. What matrix is that?

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,264 @@
+     * @return an array of length 8 in which each pair is the X and Y velocities of each wheel
+     */
+    public static double[] calculateWheelVelocities(double[] robotHeading) {
+
+
+        // multiplies M by the robotHeading to obtain the wheel velocities
+        double[] wheelVelocities = Utils.matrixVectorMult(M, robotHeading);
+
+        return wheelVelocities;
+
```
Too much whitespace.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,264 @@
+     * stops all the wheels
+     */
+    public void stop() {
+        for (int i = 0; i < swerveModules.length; i++) {
+            swerveModules[i].setSpeed(0);
+            swerveModules[i].stopAngleMotor();
```
Use a "for-each" loop:
```suggestion
        for (SwerveModule swerveModule : swerveModules) {
            swerveModule.setSpeed(0);
            swerveModule.stopAngleMotor();
```

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,264 @@
+//                M[i][2] = Rx * signX[i/2];
+//            } else {
+//                M[i][0] = 0;
+//                M[i][1] = 1;
+//                M[i][2] = Ry * signY[i/2];
+//            }
+//        }
+//        for(int i = 0; i < 8; i ++){
+//            for(int j = 0; j < 3; j++) {
+//                System.out.print(M[i][j] + " ");
+//            }
+//            System.out.println(" ");
+//        }
+//
+//        double[] robotHeading = getRobotHeading(0, 0, 1, 0);
+//
+//        double[] velocities = calculateWheelVelocities(robotHeading);
+//        double[] polar;
+//        double[][] controls = new double[4][2];
+//
+//        for(int i = 0; i <8; i++){
+//            System.out.println(velocities[i]);
+//        }
+//
+//        // converts the cartesian velocities to polar and transfers them to a control matrix
+//        for (int i = 0; i < 4; i ++) {
+//            polar = Utils.cartesianToPolar(velocities[2*i+1], velocities[2*i]);
+//            System.out.println("(" + polar[0] + ", " + polar[1] + ")");
+//            controls[i][0] = polar[0];
+//            controls[i][1] = polar[1];
+//        }
+//
+//        for (int k = 0; k < 4; k++) {
+//            System.out.println(k + " angle " + controls[k][1]);
+//            System.out.println(k + " velocity " + controls[k][0]);
+//
+//        }
+//    }
+
+}
```
Don't commit commented code.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,196 @@
+import frc.robot.Utils;
+
+public class SwerveModule extends SubsystemBase {
+    private SupplyCurrentLimitConfiguration currLimitConfig = new SupplyCurrentLimitConfiguration(true, Constants.Drivetrain.MAX_CURRENT, 5, 0.02);
```
Constants have their own file.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,196 @@
+
+        // set voltage compensation and saturation
+        driveMotor.enableVoltageCompensation(true);
+        driveMotor.configVoltageCompSaturation(12);
+
+        angleMotor.enableVoltageCompensation(true);
+        angleMotor.configVoltageCompSaturation(12);
```
Constants have their own file.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,196 @@
+                targetAngle = target;
+            }
+        }
+
```
Redundant newline.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,196 @@
+    public void periodic() {
+        // set PIDF - angle motor
+        if (wheel != 2) {
+            angleMotor.config_kP(0, Constants.SwerveModule.KP.get(), Constants.TALON_TIMEOUT);
+            angleMotor.config_kI(0, Constants.SwerveModule.KI.get(), Constants.TALON_TIMEOUT);
+            angleMotor.config_kD(0, Constants.SwerveModule.KD.get(), Constants.TALON_TIMEOUT);
+            angleMotor.config_kF(0, Constants.SwerveModule.KF.get(), Constants.TALON_TIMEOUT);
+        } else {
+            angleMotor.config_kP(0, Constants.SwerveModule.KP_SICK.get(), Constants.TALON_TIMEOUT);
+            angleMotor.config_kI(0, Constants.SwerveModule.KI_SICK.get(), Constants.TALON_TIMEOUT);
+            angleMotor.config_kD(0, Constants.SwerveModule.KD_SICK.get(), Constants.TALON_TIMEOUT);
+            angleMotor.config_kF(0, Constants.SwerveModule.KF_SICK.get(), Constants.TALON_TIMEOUT);
+        }
+
+        // set PIDF - drive motor
+        if (wheel != 1) {
+            driveMotor.config_kP(1, Constants.SwerveModule.KP_DRIVE.get(), Constants.TALON_TIMEOUT);
+            driveMotor.config_kI(1, Constants.SwerveModule.KI_DRIVE.get(), Constants.TALON_TIMEOUT);
+            driveMotor.config_kD(1, Constants.SwerveModule.KD_DRIVE.get(), Constants.TALON_TIMEOUT);
+            driveMotor.config_kF(1, Constants.SwerveModule.KF_DRIVE.get(), Constants.TALON_TIMEOUT);
+        } else {
+            driveMotor.config_kP(1, Constants.SwerveModule.KP_DRIVE_SLOW.get(), Constants.TALON_TIMEOUT);
+            driveMotor.config_kI(1, Constants.SwerveModule.KI_DRIVE_SLOW.get(), Constants.TALON_TIMEOUT);
+            driveMotor.config_kD(1, Constants.SwerveModule.KD_DRIVE_SLOW.get(), Constants.TALON_TIMEOUT);
+            driveMotor.config_kF(1, Constants.SwerveModule.KF_DRIVE_SLOW.get(), Constants.TALON_TIMEOUT);
```
You are repeating a lot of code here and in lines 44-65. Refactor your code.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/UnitModel.java:
```diff
@@ -0,0 +1,46 @@
+package frc.robot.subsystems.drivetrain;
```
Why is `UnitModel` stored inside the `drivetrain` package?

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/commands/HolonomicDrive.java:
```diff
@@ -0,0 +1,70 @@
+        forward = joystickDeadband(forward);
+        strafe = joystickDeadband(strafe) ;
+        rotation = joystickDeadband(rotation);
+//        if (forward != 0 || strafe != 0 || rotation != 0){
+//            swerveDrive.holonomicDrive(forward, strafe, rotation);
+//        }else {
+//            for(int i = 0; i < 4; i++ ){
+//                swerveDrive.swerveModules[i].setSpeed(0);
+//                swerveDrive.swerveModules[i].setAngle(swerveDrive.swerveModules[i].getAngle());
+//            }
+//        }
```
Don't commit commented code.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/commands/ResetPositions.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public void initialize() {
+        System.out.println("aaaaaa");
```
?

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TankDrive.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Override
+    public void initialize() {
+        for (int i = 0; i < 4; i++) {
+            swerveDrive.swerveModules[i].setAngle(0);
+        }
```
Use a "for-each" loop.

katzuv reviewed (changes requested):
Good night 🙂

adamv17 committed:
Utilize for each

adamv17 committed:
Use hypot

adamv17 committed:
Delete unnecessary code

adamv17 committed:
Refactor name

orel-david committed:
Update gradle dependencies

orel-david committed:
Fix getVelocity method

orel-david commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,77 @@
+    public static class Drivetrain {
+        public static final double TICKS_PER_METER = 2048 / (4 * 0.0254 * Math.PI) * 7.5;
+        public static final double TICKS_PER_RAD = 1024 / (2 * Math.PI);
+        public static final int TICKS_IN_ENCODER = 1024;
```
Use `TICKS_IN_ENCODER` in `TICKS_PER_RAD`

orel-david commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,77 @@
+        public static final double TICKS_PER_RAD = 1024 / (2 * Math.PI);
+        public static final int TICKS_IN_ENCODER = 1024;
+
+        public static final double MAX_VEL = 3;// in m/s
```
4 m/s actually

orel-david commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,77 @@
+        // the rotational speed of the robot, this constant multiplies the rotation output of the joystick
+        public static final double ROTATION_MULTIPLIER = 1.5 * Math.PI;
+
+        public static final double JOYSTICK_THRESHOLD = 0.1; // TODO: tune to actual value
```
Remove the TODO

orel-david commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,77 @@
-public final class Constants {
+public class Constants {
+    //All general constants go here
+    //public static final double TIME_STEP = 0.02;
```
Delete this line

orel-david commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,77 @@
+        public static final double JOYSTICK_THRESHOLD = 0.1; // TODO: tune to actual value
+
+        // encoder 0 for angle motors
+        public static final int[] MECHANICAL_OFFSET = new int[]{-343, -0, -0, 0}; //TODO: incorrect
```
Remove the TODO

orel-david commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,77 @@
+    }
+
+
+    public static class ExampleSubsystem1 {
+        //All of the Subsystem specific constants go here,and need to be static.
+
+        //public static final double TICKS_PER_METER = 256 / (4*0.0254*Math.PI);
+        //public static final double MAX_VELOCITY = 5;
+    }
```
Delete this

orel-david commented on src/main/java/frc/robot/OI.java:
```diff
@@ -0,0 +1,29 @@
+
+public class OI {
+
+    public static Joystick joystick = new Joystick(0);
```
Which joystick

orel-david commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -1,82 +1,91 @@
+    SendableChooser<Command> m_chooser = new SendableChooser<>();
+
+    public static XboxController xbox = new XboxController(2);
+    public static Button b = new JoystickButton(xbox, 1);
+    Button x = new JoystickButton(xbox, 3);
+    Button y = new JoystickButton(xbox, 4);
+    private JoystickButton a = new JoystickButton(xbox, XboxController.Button.kA.value);
+    private JoystickButton c = new JoystickButton(OI.joystick, 3);
```
Use only JoystickButton and put everythinh as private static

orel-david commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -1,82 +1,91 @@
+    public RobotContainer(){
+        configureButtonBindings();
+
+        //m_chooser.addOption("Example Auto 1", new DriveStraight());
+        //m_chooser.addOption("Example Auto 2", new ExampleCommand());
+        //m_chooser.setDefaultOption();
```
Delete comments

orel-david commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -1,82 +1,91 @@
+    private JoystickButton c = new JoystickButton(OI.joystick, 3);
+
+    public SwerveDrive swerveDrive = new SwerveDrive(true);
+//    public HolonomicDrive holonomicDrive = new HolonomicDrive(swerveDrive);
```
Delete comments

orel-david commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -1,82 +1,91 @@
+
+    private void configureButtonBindings() {
+        // Grab the hatch when the 'A' button is pressed.
+//        a.whenPressed(new HolonomicDrive(swerveDrive));
+//        swerveDrive.setDefaultCommand(new TurnInPlace(swerveDrive));
+//        swerveDrive.setDefaultCommand(new DriveForward(swerveDrive));
+        swerveDrive.setDefaultCommand(new HolonomicDrive(swerveDrive));
+//        swerveDrive.setDefaultCommand(new DriveForward(swerveDrive));
+//        swerveDrive.setDefaultCommand(new TankDrive(swerveDrive));
+        c.whenPressed(new ResetPositions(swerveDrive));
+        //new JoystickButton(m_driverController, Button.kB.value).whenPressed(new ExampleCommand());
```
Delete comments

orel-david commented on src/main/java/frc/robot/subsystems/drivetrain/commands/DriveForward.java:
```diff
@@ -0,0 +1,71 @@
+
+        forward = joystickDeadband(forward);
+
+        System.out.println(forward);
```
Remove this print statement

orel-david commented on src/main/java/frc/robot/subsystems/drivetrain/commands/DriveForward.java:
```diff
@@ -0,0 +1,71 @@
+    @Override
+    public void execute() {
+        double forward = -OI.getJoystickY();
+
+
```
Redundant lines

orel-david commented on src/main/java/frc/robot/subsystems/drivetrain/commands/DriveForward.java:
```diff
@@ -0,0 +1,71 @@
+
+        System.out.println(forward);
+
+//        swerveDrive.holonomicDrive(forward, 0, 0);
```
Delete this commnet

orel-david commented on src/main/java/frc/robot/subsystems/drivetrain/commands/DriveForward.java:
```diff
@@ -0,0 +1,71 @@
+        for (int i = 0; i< 4; i++) {
+            swerveDrive.swerveModules[i].setAngle(swerveDrive.swerveModules[i].getAngle());
+            swerveDrive.swerveModules[i].setSpeed(target.get());
+            System.out.println(i + " " + swerveDrive.swerveModules[i].getSpeed());
```
Remove this print statement

orel-david commented on src/main/java/frc/robot/subsystems/drivetrain/commands/HolonomicDrive.java:
```diff
@@ -0,0 +1,70 @@
+        double strafe = -RobotContainer.xbox.getX(GenericHID.Hand.kRight);
+        double rotation = -RobotContainer.xbox.getX(GenericHID.Hand.kLeft);
+        forward = joystickDeadband(forward);
+        strafe = joystickDeadband(strafe) ;
```
```suggestion
        strafe = joystickDeadband(strafe);
```

orel-david commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TankDrive.java:
```diff
@@ -0,0 +1,33 @@
+import frc.robot.subsystems.drivetrain.SwerveDrive;
+import frc.robot.subsystems.drivetrain.SwerveModule;
+
+public class TankDrive extends CommandBase {
```
Great job

adamv17 replied:
I'm keeping this

katzuv replied:
Where's the noun in the command's name?

orel-david replied:
You should

orel-david commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TurnInPlace.java:
```diff
@@ -0,0 +1,78 @@
+
+        rotation = joystickDeadband(rotation);
+
+        System.out.println(rotation);
```
Delete this print statement

orel-david commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TurnInPlace.java:
```diff
@@ -0,0 +1,78 @@
+        swerveDrive.swerveModules[2].setAngle(target2.get());
+        swerveDrive.swerveModules[3].setAngle(target3.get());
+
+        for (int i = 0; i < 4; i++){
+            System.out.println(i + " " + swerveDrive.swerveModules[i].getAngle());
+
+        }
+//        swerveDrive.holonomicDrive(0, 0, rotation);
+//        FireLog.log("target angle", target.get());
```
Delete this

Se-gol committed:
Naming fixes and auto climb fix.

adamv17 committed:
Clean up the code

adamv17 committed:
Add constants

katzuv commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,70 @@
+        public static final int[] ZERO_POSITION = {-339, 686, 221, -903};
+
+        // sick man
+        public static final WebConstant KP_SICK = new WebConstant("KP_SICK", 6);
```
Did the PID gains get infected with COVID-19?

adamv17 replied:
It got infected with engineering, so yes

katzuv replied:
🙂 
But seriously, what does "sick" man? Like driving really, really fast?

adamv17 replied:
It is the module that is not as fast as the others

katzuv replied:
Then add at least some documentation about that.

katzuv replied:
Then add at least some documentation about that.

barel13 committed:
Add documentation to the commands

adamv17 committed:
Add documentation

barel13 committed:
Merge branch 'master' into shooter

barel13 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,70 @@
+        public static final double TICKS_PER_RAD = TICKS_IN_ENCODER / (2 * Math.PI);
+
+        public static final double MAX_VEL = 4;// in m/s
+        public static final double TIME_STEP = 0.02; // in seconds
```
Timestamp between what

katzuv replied:
@barel13 used `LOOP_PERIOD` in his code, I think this wording is better, WPILib Docs use "loop period".

barel13 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,70 @@
+        public static final int TICKS_IN_ENCODER = 1024;
+        public static final double TICKS_PER_RAD = TICKS_IN_ENCODER / (2 * Math.PI);
+
+        public static final double MAX_VEL = 4;// in m/s
```
```suggestion
        public static final double MAX_VELOCITY = 4;// in m/s
```

barel13 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,70 @@
+
+        public static final boolean ENABLE_CURRENT_LIMIT = true;
+        public static final boolean ENABLE_VOLTAGE_COMPENSATION = true;
+        public static final int VOLTAGE_SATURATION = 12;
```
I think this constant should be the the robot constants, it's not specific to the subsystem

barel13 commented on src/main/java/frc/robot/OI.java:
```diff
@@ -0,0 +1,29 @@
+
+public class OI {
+
+    public static Joystick rightJoystick = new Joystick(0);
```
```suggestion
    private static final Joystick rightJoystick = new Joystick(0);
```

barel13 commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -1,82 +1,77 @@
+    public static XboxController xbox = new XboxController(2);
+    private static Button b = new JoystickButton(xbox, 1);
+    private static JoystickButton a = new JoystickButton(xbox, XboxController.Button.kA.value);
+    private static JoystickButton c = new JoystickButton(OI.rightJoystick, 3);
```
Move it to OI class

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,221 @@
+
+public class SwerveDrive extends SubsystemBase {
+
+    static double[][] Dynamics = new double[8][3];
```
```suggestion
    static double[][] dynamics = new double[8][3];
```

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,221 @@
+        swerveModules[0] = new SwerveModule(0, new TalonFX(DRIVE_MOTOR_1), new TalonSRX(ANGLE_MOTOR_1), FRONT_RIGHT_INVERTED);
+        swerveModules[1] = new SwerveModule(1, new TalonFX(DRIVE_MOTOR_2), new TalonSRX(ANGLE_MOTOR_2), FRONT_LEFT_INVERTED);
+        swerveModules[2] = new SwerveModule(2, new TalonFX(DRIVE_MOTOR_3), new TalonSRX(ANGLE_MOTOR_3), BACK_RIGHT_INVERTED);
+        swerveModules[3] = new SwerveModule(3, new TalonFX(DRIVE_MOTOR_4), new TalonSRX(ANGLE_MOTOR_4), BACK_LEFT_INVERTED);
```
if the number of the module is `0,1,2,3`, don't you should apply the same conventions to the Constants?

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,221 @@
+            swerveModules[k].setAngle(controls[k][1]);
+        }
+
+        double sumx = 0;
+        double sumy = 0;
+        for (int j = 0; j < 4; j++) {
+            sumx += velocities[j * 2];
+            sumy += velocities[j * 2 + 1];
+        }
+        double[] target = Utils.cartesianToPolar(sumx, sumy);
+        FireLog.log("target velocity", target[0]);
+        FireLog.log("target angle", target[1]);
```
I would move it to be inside if statement that checks whether the the robot in debug mode

orel-david replied:
The firelog should only work if the robot is in debug mode so it is OK

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,183 @@
+            Constants.SwerveModule.TRIGGER_THRESHOLD_CURRENT,
+            Constants.SwerveModule.TRIGGER_THRESHOLD_TIME
+    );
+    public final TalonFX driveMotor;
```
pubilc?

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,183 @@
+
+    public SwerveModule(int wheel, TalonFX driveMotor, TalonSRX angleMotor, boolean[] inverted) {
+        // configure feedback sensors
+        angleMotor.configSelectedFeedbackSensor(FeedbackDevice.Analog, wheel, Constants.TALON_TIMEOUT);
```
```suggestion
        angleMotor.configSelectedFeedbackSensor(FeedbackDevice.Analog, 0, Constants.TALON_TIMEOUT);
```
?

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/DriveForward.java:
```diff
@@ -0,0 +1,65 @@
+public class DriveForward extends CommandBase {
+
+    private SwerveDrive swerveDrive;
+    private WebConstant target = new WebConstant("targetSpeed", 0);
```
```suggestion
    private WebConstant targetSpeed = new WebConstant("targetSpeed", 0);
```

orel-david replied:
I don't think we need the word speed because the command is drive forward

barel13 replied:
I only suggested it because the name of the WebConstant is targetSpeed

orel-david replied:
We can change the name of the Webconstant

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/DriveForward.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    public void execute() {
+        double forward = -OI.getJoystickY();
+        forward = joystickDeadband(forward);
```
You can combine this into 1 line
```suggestion
        double forward = joystickDeadband(-OI.getJoystickY());
```

natymeitav committed:
Rename StartFunnel's parameters in constructor.

Co-authored-by: orel-david <42713127+orel-david@users.noreply.github.com>

natymeitav committed:
Merge InTake & outTake to one command.

natymeitav committed:
Add option to move funnel's motor in 2 directions.

natymeitav committed:
Assign a button to move balls to the intake.

natymeitav committed:
Set motor to be VictorSPX type.

natymeitav committed:
Make Funnel's motors work based of power instead of velocity.

barel13 committed:
Apply conventions to the constant class

barel13 committed:
Remove empty methods

barel13 committed:
Rename rpsUnitModel to unitModel

barel13 committed:
Fix documentation in Shooter class

barel13 committed:
"Disable" estimateVelocity function

barel13 committed:
Save getVelocity to a variable

barel13 committed:
Remove unnecessary condition

barel13 committed:
Clarify the documentation of the Shoot command

barel13 committed:
Update the commands to handle a case when the velocity changes

barel13 committed:
Fix the units of OMEGA constant

Co-authored-by: Dan Katzuv <31829093+katzuv@users.noreply.github.com>

barel13 committed:
Fix SpeedUp velocity variable use

barel13 committed:
Merge remote-tracking branch 'origin/shooter' into shooter

barel13 committed:
Initialize default classes for the conveyor

barel13 committed:
Rename isReady function to hasReachedSetpoint

barel13 committed:
Create and Implement Moving Average class

barel13 committed:
Use Moving Average object in the Shooter subsystem

barel13 committed:
Remove .gradle files

barel13 committed:
Ctrl+Alt+L on Robot.java file

barel13 committed:
Fix Moving Average

barel13 committed:
Create another constructor to MovingAverage class

barel13 committed:
Create example of Shooting data

barel13 committed:
Bind the path to a constant

barel13 committed:
Create example of how to read the file inside the Shooter class

barel13 committed:
Fix names

barel13 committed:
Apply conventions on MovingAverage class

barel13 committed:
Create And implement DeadbandProximity class

barel13 committed:
Create Conveyor Motor

barel13 committed:
Create function for the conveyor

barel13 committed:
Add method to handle the balls

barel13 committed:
Create LoadConveyor command

barel13 committed:
Add safety feature to LoadConveyor command

barel13 committed:
Ignore .idea folder

barel13 committed:
Rename LoadConveyor command to Convey

barel13 committed:
Add documentation to the conveyor subsystem

barel13 committed:
Add support to a case when the conveyor take out ball threw the funnel

barel13 committed:
Remove the UnitModel from the conveyor

barel13 committed:
Fix the convey if statement

barel13 committed:
Add boolean to check the whether we shoot or store the balls

barel13 committed:
Fix build issue

barel13 committed:
Remove Utils class

barel13 committed:
Remove unnecessary code, and fix documentation

barel13 committed:
Rename moving average variable

barel13 committed:
Fix documentation

barel13 committed:
Split constructor to multiple methods

barel13 committed:
Clarify comment

barel13 committed:
Retrieve the distances from getClosestDistances

barel13 committed:
Make unreachable catch more "generic"

barel13 committed:
Rename ROBOT_TIMEOUT constant

barel13 committed:
Change the purpose of the getPower method

barel13 committed:
make lastState boolean final

barel13 committed:
Add braced to convey command, and add TODO

barel13 committed:
Use constants instead of "magic numbers"

barel13 committed:
Fix documentation

barel13 committed:
Change isMovingUp logic

katzuv commented on src/main/java/frc/robot/OI.java:
```diff
@@ -0,0 +1,29 @@
+
+import edu.wpi.first.wpilibj.Joystick;
+
+public class OI {
```
Move this class to another PR that adds all things related to OI and joysticks.

barel13 committed:
Fix documentation and constants

barel13 committed:
Change the logic inside MovingAverage

barel13 committed:
Change the check for the distances equality in MovingAverage

barel13 committed:
Remove if statement checking

barel13 committed:
Remove SpeedUp command

barel13 committed:
Add difference to setVelocity function

barel13 committed:
Apply the changes to Shoot command

orel-david commented on src/main/java/frc/robot/subsystems/drivetrain/commands/DriveForward.java:
```diff
@@ -0,0 +1,65 @@
+        FireLog.log("swerve direction", Robot.gyro.getAngle());
+    }
+
+    /**
+     * sets the value of the joystick to 0 if the value is less than the threshold
+     *
+     * @param val the joystick value
+     * @return 0 if val is less than the threshold else val
+     */
+    private double joystickDeadband(double val) {
+        if (val < Constants.SwerveDrive.JOYSTICK_THRESHOLD)
+            return 0;
+        return val;
+    }
```
Maybe you should move it to the utils file

orel-david commented on src/test/java/robot/subsystems/drivetrain/SwerveDriveTest.java:
```diff
@@ -0,0 +1,132 @@
+
+import static frc.robot.Constants.SwerveDrive.*;
+
+public class SwerveDriveTest {
```
Notice that 4 of your tests are failing probably due to the radius not being 1 meter

adamv17 committed:
Move joysticks and buttons to OI

adamv17 committed:
Combine

adamv17 committed:
Refactor names

adamv17 committed:
Move to Utils

adamv17 committed:
Adjust test

adamv17 committed:
Delete line that crashes the code

barel13 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,69 @@
-public final class Constants {
+public class Constants {
+    public static final int TALON_TIMEOUT = 10; // in ms
+    public static final boolean ENABLE_CURRENT_LIMIT = true;
```
is it related to the whole robot?

katzuv replied:
That's actually a good question… do you want to apply current limit on all motors?

barel13 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,69 @@
+    public static final int TALON_TIMEOUT = 10; // in ms
+    public static final boolean ENABLE_CURRENT_LIMIT = true;
+    public static final boolean ENABLE_VOLTAGE_COMPENSATION = true;
+    public static final int VOLTAGE_SATURATION = 12;
```
units

barel13 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -1,19 +1,69 @@
+
+        public static final double MAX_VELOCITY = 4;// in m/s
+        public static final double TIME_STEP = 0.02; // in seconds, time step of run time
+        public static final double MAX_ACCELERATION = 0.4;// in m/s^2 (currently not the correct number)
```
You can remove this constants, you don't use them

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,183 @@
+import frc.robot.Utils;
+
+public class SwerveModule extends SubsystemBase {
+    private SupplyCurrentLimitConfiguration currLimitConfig = new SupplyCurrentLimitConfiguration(
```
You can make it local

adamv17 replied:
make what local?

barel13 replied:
`currLimitConfig` you only use it inside the constructor.

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,183 @@
+    private final int wheel;
+
+    private UnitModel unitDrive = new UnitModel(Constants.SwerveDrive.TICKS_PER_METER);
+    private UnitModel unitAngle = new UnitModel(Constants.SwerveDrive.TICKS_PER_RAD);
```
```suggestion
    private final UnitModel unitDrive = new UnitModel(Constants.SwerveDrive.TICKS_PER_METER);
    private final UnitModel unitAngle = new UnitModel(Constants.SwerveDrive.TICKS_PER_RAD);
```

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,183 @@
+        angleMotor.configVoltageCompSaturation(Constants.VOLTAGE_SATURATION);
+
+        angleMotor.selectProfileSlot(0, 0);
+        driveMotor.selectProfileSlot(1, 0);
```
why the slot index is 1? (it's not the same motor)

adamv17 replied:
There was a bug related to this, it works and I don't want to change it

katzuv replied:
That sounds pretty weird. What was the bug?
I recommend checking again because now you need to use that `1` everywhere. If there really is a bug, report it to CTRE and benefit the whole FRC community 🙂.

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,183 @@
+    private UnitModel unitDrive = new UnitModel(Constants.SwerveDrive.TICKS_PER_METER);
+    private UnitModel unitAngle = new UnitModel(Constants.SwerveDrive.TICKS_PER_RAD);
+
+    public SwerveModule(int wheel, TalonFX driveMotor, TalonSRX angleMotor, boolean[] inverted) {
```
I would pass the port of the angle and the drive motors instead of the actual motor object

adamv17 replied:
What are the upsides to this?

adamv17 replied:
actually nevermind I see that it will help fix the tests

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,183 @@
+     * @return the speed of the wheel in [m/s]
+     */
+    public double getSpeed() {
+        return unitDrive.toVelocity(driveMotor.getSelectedSensorVelocity(1));
```
Are you sure it should be 1?

adamv17 replied:
yes

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,183 @@
+     */
+    public void setAngle(double angle) {
+        double targetAngle = getTargetAngle(angle, getAngle());
+        //int targetTicks = getTargetTicks(angle);
```
Remove this line

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,183 @@
+        angleMotor.setSelectedSensorPosition(0);
+    }
+
+    /**
+     * set the drive power
+     *
+     * @param power a value between [-1, 1]
+     */
+    public void setPower(double power) {
+        driveMotor.set(ControlMode.PercentOutput, power);
+    }
```
you don't use this function, and if you are going to use it you need to rename the function, its unclear whether you apply power on the drive motor or the angle motor

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/DriveForward.java:
```diff
@@ -0,0 +1,64 @@
+        addRequirements(swerveDrive);
+    }
+
+    @Override
+    public void initialize() {
+    }
```
You can remove this function

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/DriveForward.java:
```diff
@@ -0,0 +1,64 @@
+        double forward = joystickDeadband(-OI.getJoystickY());
+
+        for (int i = 0; i< 4; i++) {
+            swerveDrive.swerveModules[i].setAngle(swerveDrive.swerveModules[i].getAngle());
+            swerveDrive.swerveModules[i].setSpeed(target.get());
```
make a getter, don't pass it directly

adamv17 replied:
why

barel13 replied:
I mean a getter for the module (`swerveDrive.get(i)`...)

adamv17 replied:
got it

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/HolonomicDrive.java:
```diff
@@ -0,0 +1,53 @@
+
+public class HolonomicDrive extends CommandBase {
+
+    private SwerveDrive swerveDrive;
```
```suggestion
    private final SwerveDrive swerveDrive;
```

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/HolonomicDrive.java:
```diff
@@ -0,0 +1,53 @@
+    @Override
+    public void initialize() {
+        Robot.gyro.reset();
+
```
remove this empty line

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/HolonomicDrive.java:
```diff
@@ -0,0 +1,53 @@
+        if (forward != 0 || strafe != 0 || rotation != 0) {
+            swerveDrive.holonomicDrive(forward, strafe, rotation);
+        } else {
+            for (int i = 0; i < 4; i++) {
+                swerveDrive.swerveModules[i].setSpeed(0);
+                swerveDrive.swerveModules[i].setAngle(swerveDrive.swerveModules[i].getAngle());
+            }
```
why don't you use the `stop()` function instead?

adamv17 replied:
We want the wheels to keep their current position

barel13 replied:
and what is the difference? inside the `stop` function you stop every `driveMotor` and set the power to apply by the `angleMotor` to 0.

adamv17 replied:
here the angle motor continues to apply power to make sure the will stays on the same angle

katzuv replied:
Extract this "locking" logic into a method of `SwerveModule`, you use this logic in more than one place.

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/ResetPositions.java:
```diff
@@ -0,0 +1,20 @@
+
+public class ResetPositions extends InstantCommand {
+
+    private SwerveDrive swerveDrive;
```
```suggestion
    private final SwerveDrive swerveDrive;
```

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TurnInPlace.java:
```diff
@@ -0,0 +1,75 @@
+
+public class TurnInPlace extends CommandBase {
+
+    private SwerveDrive swerveDrive;
+    //    private WebConstant target = new WebConstant("targetAngle", 0);
+    private WebConstant target0 = new WebConstant("target0", 0);
+    private WebConstant target1 = new WebConstant("target1", 0);
+    private WebConstant target2 = new WebConstant("target2", 0);
+    private WebConstant target3 = new WebConstant("target3", 0);
```
```suggestion
    private final SwerveDrive swerveDrive;
    private final WebConstant target0 = new WebConstant("target0", 0);
    private final WebConstant target1 = new WebConstant("target1", 0);
    private final WebConstant target2 = new WebConstant("target2", 0);
    private final WebConstant target3 = new WebConstant("target3", 0);
```

and what is target0, target1, etc...

adamv17 replied:
target for each wheel numbered by index

barel13 replied:
I would appreciate it if you will add a comment that explains that

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TurnInPlace.java:
```diff
@@ -0,0 +1,75 @@
+        addRequirements(swerveDrive);
+    }
+
+    @Override
+    public void initialize() {
+    }
```
Remove this empty function

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TurnInPlace.java:
```diff
@@ -0,0 +1,75 @@
+        FireLog.log("swerve direction", Robot.gyro.getAngle());
+    }
+
+    /**
+     * sets the value of the joystick to 0 if the value is less than the threshold
+     *
+     * @param val the joystick value
+     * @return 0 if val is less than the threshold else val
+     */
+    private double joystickDeadband(double val) {
+        if (val < Constants.SwerveDrive.JOYSTICK_THRESHOLD)
+            return 0;
+        return val;
+    }
```
If you have this function all over the commands, why don't you move it to the `Utils` class?

adamv17 replied:
I did move it to the utils I forget to implement it here

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TurnInPlace.java:
```diff
@@ -0,0 +1,75 @@
+    @Override
+    public void execute() {
+        double rotation = -OI.xbox.getY();
+
```
You don't need this empty line

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TurnInPlace.java:
```diff
@@ -0,0 +1,75 @@
+
+        rotation = joystickDeadband(rotation);
+
+        swerveDrive.swerveModules[0].setAngle(target0.get());
+        swerveDrive.swerveModules[1].setAngle(target1.get());
+        swerveDrive.swerveModules[2].setAngle(target2.get());
+        swerveDrive.swerveModules[3].setAngle(target3.get());
```
don't pass directly

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TurnInPlace.java:
```diff
@@ -0,0 +1,75 @@
+        swerveDrive.swerveModules[3].setAngle(target3.get());
+
+        for (int i = 0; i < 4; i++) {
+            System.out.println(i + " " + swerveDrive.swerveModules[i].getAngle());
```
Why not using the firelog instead?

barel13 committed:
Rename DeadbandProximity to DeadbandLaserSensor

barel13 committed:
Add another port to the funnel sensor

adamv17 committed:
General fixes

barel13 committed:
Add documentation to the sensor state-machine

adamv17 committed:
Implement deadband from utils

adamv17 committed:
Pass the port instead of motor

adamv17 committed:
Create getModule method

adamv17 committed:
More fixes

adamv17 committed:
Fix TurnInPlace and driveForward tests

katzuv commented on src/main/java/frc/robot/Utils.java:
```diff
@@ -0,0 +1,82 @@
+     * @param angle the angle of the velocity vector
+     * @return the cartesian representation with x and y components
+     */
+    public static double[] polarToCartesian(double velocity, double angle) {
```
This method is a utility, so you should use a more general term than "velocity". from Wikipedia:
> The distance from the pole is called the _radial coordinate_, _radial distance_ or simply _radius_, and the angle is called the _angular coordinate_, _polar angle_, or _azimuth_. ([source](https://en.wikipedia.org/wiki/Polar_coordinate_system))

So I think "radius" and "angle" will be the most suitable terms here. General, and not too long.

katzuv commented on src/main/java/frc/robot/Utils.java:
```diff
@@ -0,0 +1,82 @@
+     * @return the cartesian representation with x and y components
+     */
+    public static double[] polarToCartesian(double velocity, double angle) {
+        double[] v = new double[2];
```
Don't shorten. `coordinate` will do fine.
The same applies to line 27:
https://github.com/Galaxia5987/robot-2021/blob/1897ddaaacf654968c1bc54acc19b7fd8232d824/src/main/java/frc/robot/Utils.java#L27

adamv17 replied:
Could you rephrase I don't understand

katzuv replied:
```suggestion
        double[] coordniate = new double[2];
```
Or:
```suggestion
        double[] cartesianCoordinate = new double[2];
```

katzuv commented on src/main/java/frc/robot/Utils.java:
```diff
@@ -0,0 +1,82 @@
+
+    /**
+     * calculates the matrix - vector multiplication
+     * assuming that the number of columns in the matrix correspond to the number of rows in the vector
```
```suggestion
     * Assuming that the number of columns in the matrix is equal to the number of rows in the vector.
```

katzuv commented on src/main/java/frc/robot/Utils.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    /**
+     * calculates the matrix - vector multiplication
```
```suggestion
     * Calculates a matrix-vector multiplication.
```

katzuv commented on src/main/java/frc/robot/Utils.java:
```diff
@@ -0,0 +1,82 @@
+     * @return 0 if val is less than the threshold else val
+     */
+    public static double joystickDeadband(double val) {
+        if (Math.abs(val) < Constants.SwerveDrive.JOYSTICK_THRESHOLD)
```
This function is in a general utility class. Therefore, it should work for every value and threshold.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+    public SwerveDrive(boolean isFieldOriented, boolean testMode) {
+
+        // creates an inverse matrix of all the mathematical operations needed to calculate the wheel velocities
+        // see https://file.tavsys.net/control/controls-engineering-in-frc.pdf pg.144
```
Are you sure that's the right page number? The book might have updated, but page 144 doesn't seem to talk about relevant stuff to these calculations.

adamv17 replied:
You are right the page is 140

katzuv replied:
So please update the documentation.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+                dynamics[i][1] = 1;
+                dynamics[i][2] = Ry * signY[i / 2];
+            }
+        }
```
Extract these lines to another method.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+                dynamics[i][2] = Ry * signY[i / 2];
+            }
+        }
+        Robot.gyro.reset();
```
This should be called in `Robot.robotInit()` or maybe in `Robot.autonomousInit()`. I think the first is better, because the reset might take a few seconds, and the reset won't finish before the robot starts moving.

katzuv commented on src/main/java/frc/robot/Ports.java:
```diff
@@ -0,0 +1,38 @@
+        // front right
+        public static final int DRIVE_MOTOR_0 = 11;
+        public static final int ANGLE_MOTOR_0 = 12;
+        public static final boolean[] FRONT_RIGHT_INVERTED = {false, false, true, false};
```
Why do you use "0" but "front right", "1" but "front left", etc.? Be consistent.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+            swerveModules[k].setAngle(controls[k][1]);
+        }
+
+        double sumx = 0;
+        double sumy = 0;
+        for (int j = 0; j < 4; j++) {
+            sumx += velocities[j * 2];
+            sumy += velocities[j * 2 + 1];
```
Could you please explain what these lines do?

adamv17 replied:
They reconstruct the current velocity of the robot based on the wheel speeds and angles

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+     */
+    public static double[] getRobotHeading(double forward, double strafe, double rotation, double robotAngle) {
+        // turns the joystick values into the heading of the robot
+        forward *= Constants.SwerveDrive.SPEED_MULTIPLIER;
+        strafe *= Constants.SwerveDrive.SPEED_MULTIPLIER;
+        rotation *= Constants.SwerveDrive.ROTATION_MULTIPLIER;
```
IMO, these multiplications should happen in the command, and not here. This method shouldn't get "joystick values", but commands for each axis. I believe you want to use this method in autonomous driving, no? If yes, you'd probably not want these multiplications (which will be a bug and not just a design issue).

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,176 @@
+public class SwerveModule extends SubsystemBase {
+    private final TalonFX driveMotor;
+    private final TalonSRX angleMotor;
+    private final int wheel;
```
I think "index" is better.

katzuv commented on src/main/java/frc/robot/Utils.java:
```diff
@@ -0,0 +1,82 @@
+            }
+            out[i] = sum;
+        }
+
```
```suggestion
```

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+            swerveModules[2] = new SwerveModule(2, DRIVE_MOTOR_2, ANGLE_MOTOR_2, BACK_RIGHT_INVERTED);
+            swerveModules[3] = new SwerveModule(3, DRIVE_MOTOR_3, ANGLE_MOTOR_3, BACK_LEFT_INVERTED);
+        }
+
```
```suggestion
```

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+
+        this.isFieldOriented = isFieldOriented;
+    }
+
```
```suggestion
```

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+        System.out.println("relative strafe" + speeds[1]);
+
+        double[] robotHeading = {forward, strafe, rotation};
+
```
```suggestion
```

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+            swerveModule.setSpeed(0);
+            swerveModule.stopAngleMotor();
+        }
+
```
```suggestion
```

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+        double[] lockAngles = new double[4];
+
+        for (int i = 0; i < 4; i++) {
+            lockAngles[i] = Math.PI / 2 - Math.atan(Constants.SwerveDrive.ROBOT_WIDTH / Constants.SwerveDrive.ROBOT_LENGTH) + i * Math.PI / 2;
```
Expand this line to a few variables, it's too long.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+
+        for (int i = 0; i < 4; i++) {
+            swerveModules[i].setSpeed(0);
+            if (i == 2) {
+                swerveModules[i].setAngle(lockAngles[i + 1]);
+            }
+            if (i == 3) {
+                swerveModules[i].setAngle(lockAngles[i - 1]);
```
Could you please explain this workaround?

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+     * set the angle of the wheels on the robot to lock the robot in place
+     */
+    public void lock() {
+        // calculates the lock angles of the wheels
```
This comment is redundant, it doesn't offer additional information.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+    }
+
+    /**
+     * set the angle of the wheels on the robot to lock the robot in place
```
Can you explain what that means?

adamv17 replied:
Lock angles are used for when there is defence on the robot, the wheels point outwards and it is nearly impossible to move the robot.

katzuv replied:
Oh, that's cool. Please elaborate on that in the documentation, so the next confused person won't need to ask you.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,176 @@
+
+    @Override
+    public void periodic() {
+        configPIDF();
```
Why does that need to be called every loop? Can the wheel number change mid-match?

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/commands/HolonomicDrive.java:
```diff
@@ -0,0 +1,52 @@
+
+import edu.wpi.first.wpilibj.GenericHID;
+import edu.wpi.first.wpilibj2.command.CommandBase;
+import frc.robot.*;
```
Please don't. The `robot` package is huge, you're cluttering your namespace with many classes and functions you don't need.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/commands/DriveForward.java:
```diff
@@ -0,0 +1,49 @@
+        double forward = Utils.joystickDeadband(-OI.getJoystickY());
+
+        for (int i = 0; i< 4; i++) {
+            swerveDrive.getModule(i).setAngle(swerveDrive.swerveModules[i].getAngle());
```
I might be misunderstanding here something, but shouldn't all modules be pointed in the same direction?
Also, if you do `set module angle to module angle`, you can just remove this line, as it does nothing.

adamv17 replied:
Ideally they should be, but at high speeds the wheels may drift and without this stablizing them there could be a problem.

katzuv replied:
If you call `talon.set`, the motor controller's setpoint will remain the same until you call `set` again. So if you called `angleMotor.set(ControlMode.Position, angleTicks)` and didn't change the setpoint, the motor controller will still try to minimize the error, thus achieving the stabilization you're looking for.
> they should be

So I think a method of `SwerveDrive` that sets the angle for all modules will be nice.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/commands/HolonomicDrive.java:
```diff
@@ -0,0 +1,52 @@
+        double forward = Utils.joystickDeadband(-OI.xbox.getY(GenericHID.Hand.kRight));
+        double strafe = Utils.joystickDeadband(-OI.xbox.getX(GenericHID.Hand.kRight));
+        double rotation = Utils.joystickDeadband(-OI.xbox.getX(GenericHID.Hand.kLeft));
+        if (forward != 0 || strafe != 0 || rotation != 0) {
```
Why is this condition needed?

adamv17 replied:
If I remember correctly we had a problem where the robot would move even if the joystick values were zero. @orel-david is that correct?

orel-david replied:
Sometimes when we released the joysticks the modules returned to the origin when they could just stop instead

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,176 @@
+    private final TalonSRX angleMotor;
+    private final int wheel;
+
+    private final UnitModel unitDrive = new UnitModel(Constants.SwerveDrive.TICKS_PER_METER);
+    private final UnitModel unitAngle = new UnitModel(Constants.SwerveDrive.TICKS_PER_RAD);
```
Please improve the names of these converters.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,176 @@
+     * @return the angle of the wheel in radians
+     */
+    public double getAngle() {
+        return unitAngle.toUnits(angleMotor.getSelectedSensorPosition() - Constants.SwerveModule.ZERO_POSITION[wheel]);
```
> `- Constants.SwerveModule.ZERO_POSITION[wheel]`

If I understand the intention correctly, you want to compensate for the encoder offset. Won't it be simpler to call `angleMotor.setSelectedSensorPosition()` with the "zero position" in this class' constructor?

adamv17 replied:
Yes it would be

katzuv commented on build.gradle:
```diff
@@ -67,13 +67,14 @@ dependencies {
 
     implementation 'com.sparkjava:spark-core:2.8.0'
     implementation 'com.sparkjava:spark-template-velocity:2.7.1'
+    implementation "com.revrobotics.frc:SparkMax-java:1.5.2"
```
I don't think you need this line, but in any case, where are you using REV products?

adamv17 replied:
It was for the NEO in the intake

katzuv replied:
So it shouldn't be in this PR anyway.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TurnInPlace.java:
```diff
@@ -0,0 +1,52 @@
+import frc.robot.valuetuner.WebConstant;
+import org.techfire225.webapp.FireLog;
+
+public class TurnInPlace extends CommandBase {
```
"rotate" might be a more accurate term.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/commands/TankDrive.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public void execute() {
+        swerveDrive.getModule(0).setSpeed(OI.xbox.getY(GenericHID.Hand.kRight));
+        swerveDrive.getModule(2).setSpeed(OI.xbox.getY(GenericHID.Hand.kRight));
+
+        swerveDrive.getModule(1).setSpeed(OI.xbox.getY(GenericHID.Hand.kLeft));
+        swerveDrive.getModule(3).setSpeed(OI.xbox.getY(GenericHID.Hand.kLeft));
```
Extract `OI.xbox.getY(GenericHID.Hand.kRight)` and `OI.xbox.getY(GenericHID.Hand.kLeft)` to variables, to reduce function calls.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/commands/HolonomicDrive.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public void execute() {
+        double forward = Utils.joystickDeadband(-OI.xbox.getY(GenericHID.Hand.kRight));
+        double strafe = Utils.joystickDeadband(-OI.xbox.getX(GenericHID.Hand.kRight));
+        double rotation = Utils.joystickDeadband(-OI.xbox.getX(GenericHID.Hand.kLeft));
```
How will this command used in autonomous driving?

katzuv reviewed (changes requested):
WPILib has linear algebra tools, which are based on EJML ("Efficient Java Matrix Library"). I guess you saw from tests that the code is efficient enough, but if needed, WPILib might be helpful.

barel13 committed:
Update gradle version

until I updated gradle, I couldn't compile the code...

barel13 committed:
Rename estimateVelocity to evaluateVelocity

barel13 committed:
Add CANSparkMax dependency

barel13 committed:
Refactor Convey to FeedShooter

barel13 committed:
Create LoadConveyor command

barel13 committed:
Add methods to handle more appropriately the use of the balls

barel13 committed:
Implement LoadConveyor command

barel13 committed:
Add documentation to LoadConveyor command

barel13 committed:
Fix documentation

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,176 @@
+            driveMotor.config_kD(1, Constants.SwerveModule.KD_DRIVE_SLOW.get(), Constants.TALON_TIMEOUT);
+            driveMotor.config_kF(1, Constants.SwerveModule.KF_DRIVE_SLOW.get(), Constants.TALON_TIMEOUT);
+        }
+    }
```
Consider simply passing to this class's constructor the needed PIDF gains.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+
+        double[] robotHeading = {forward, strafe, rotation};
+
+        return robotHeading;
```
Redundant variable.
```suggestion
        return new double[]{forward, strafe, rotation};
```

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+    public static double[] calculateWheelVelocities(double[] robotHeading) {
+        // multiplies M by the robotHeading to obtain the wheel velocities
+        double[] wheelVelocities = Utils.matrixVectorMult(dynamics, robotHeading);
+        return wheelVelocities;
```
Redundant variable.
```suggestion
        return Utils.matrixVectorMult(dynamics, robotHeading);
```

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/commands/ResetPositions.java:
```diff
@@ -0,0 +1,20 @@
+import edu.wpi.first.wpilibj2.command.InstantCommand;
+import frc.robot.subsystems.drivetrain.SwerveDrive;
+
+public class ResetPositions extends InstantCommand {
```
Consider using an [inlined `InstantCommand`](https://docs.wpilib.org/en/stable/docs/software/commandbased/convenience-features.html#instantcommand):
```java
new InstantCommand(swerveDrive::resetAll, swerveDrive);
```

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+    }
+
+    /**
+     * resets all the modules
```
This is confusing, I thought that sets the setpoint of all the modules to zero, not the encoders values.

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+    /**
+     * resets all the modules
+     */
+    public void resetAll() {
```
```suggestion
    public void resetAllEncoders() {
```

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,176 @@
+    /**
+     * reset encoder value to 0
+     */
+    public void resetAngle() {
```
```suggestion
    public void resetAngleEncoder() {
```

katzuv commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,231 @@
+    public void resetAll() {
+        for (int i = 0; i < 4; i++) {
+            swerveModules[i].resetAngle();
+        }
```
```suggestion
        for (SwerveModule swerveModule : swerveModules) {
            swerveModule.resetAngle();
        }
```

katzuv reviewed (changes requested):
There are still some unregarded comments, please go over them 🙂.
Also, please go to IntelliJ IDEA and run "Inspect code" (press <kbd>Shift</kbd> twice, then search for "inspect", or <kbd>Alt</kbd> + <kbd>Z</kbd>, then <kbd>I</kbd>). There are many unused `import`s, variables which can be declared final, etc.

barel13 committed:
Add todo to the constant of the state-space model

barel13 committed:
Rename evaluateVelocity function

barel13 committed:
Fix documentation

barel13 committed:
Rename evaluateVelocity function

barel13 committed:
Rename MovingAverage to LinearRegression

barel13 committed:
Create setVelocity wrapper function

barel13 committed:
Add units to ARBITRARY_FEED_FORWARD constant

barel13 committed:
Use AtomicInteger instead of int

Just to avoid the case that we call some related function in parallel when we change the amount of balls, for example, we check whether the conveyor is full and we have 4 balls, and by the same time we insert another ball, what would be the value from the `isConveyorFull()`

Se-gol committed:
Fixing Climber

barel13 committed:
Single-thread

adamv17 committed:
General fixes

adamv17 committed:
Add helper methods and refactor

adamv17 committed:
Document

adamv17 committed:
Pass PIDF through constructor

adamv17 committed:
Fix tests

adamv17 committed:
Optimize imports and fix error

adamv17 committed:
Delete spaces

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,251 @@
+
+import static frc.robot.Ports.SwerveDrive.*;
+
+public class SwerveDrive extends SubsystemBase {
```
Add documentation

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,181 @@
+    private final TalonFX driveMotor;
+    private final TalonSRX angleMotor;
+
+    private double angleKP;
+    private double angleKI;
+    private double angleKD;
+    private double angleKF;
+
+    private double driveKP;
+    private double driveKI;
+    private double driveKD;
+    private double driveKF;
```
Use arrays!

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,251 @@
+public class SwerveDrive extends SubsystemBase {
+
+    static double[][] dynamics = new double[8][3];
+    public SwerveModule[] swerveModules = new SwerveModule[4];
```
make it private, you already have `getModule()` function

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,181 @@
+import frc.robot.UnitModel;
+import frc.robot.Utils;
+
+public class SwerveModule extends SubsystemBase {
```
add documentation

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveModule.java:
```diff
@@ -0,0 +1,181 @@
+     */
+    public void setAngle(double angle) {
+        double targetAngle = getTargetAngle(angle, getAngle());
+        int angleTicks = angleUnitModel.toTicks(targetAngle);
```
inline!

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/DriveForward.java:
```diff
@@ -0,0 +1,49 @@
+
+public class DriveForward extends CommandBase {
+
+    private SwerveDrive swerveDrive;
```
```suggestion
    private final SwerveDrive swerveDrive;
```

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/commands/Rotate.java:
```diff
@@ -0,0 +1,55 @@
+
+public class Rotate extends CommandBase {
+
+    private SwerveDrive swerveDrive;
```
```suggestion
    private final SwerveDrive swerveDrive;
```

barel13 commented on src/main/java/frc/robot/valuetuner/WebConstant.java:
```diff
@@ -2,6 +2,8 @@
 
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
+import static spark.Spark.post;
+
```
?

adamv17 replied:
I don't know I'm not going to touch it

barel13 commented on src/main/java/frc/robot/valuetuner/WebConstantPIDTalon.java:
```diff
@@ -4,6 +4,7 @@
 
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
+import static spark.Spark.post;
```
?

adamv17 committed:
Change to private

adamv17 committed:
Inline

Se-gol committed:
General Fixes to Climber.

adamv17 committed:
Pass PIDF arrays instead

barel13 committed:
Fix documentation

barel13 committed:
Alter FeedShooter finish condition

barel13 committed:
General fixes

adamv17 committed:
Document

barel13 committed:
Create MinimizeConveyor command

orel-david committed:
Add pto ports

orel-david committed:
Add pto subsystem

orel-david committed:
Create the subsystem command

orel-david committed:
Capitalize the enum

Se-gol committed:
Update src/main/java/frc/robot/subsystems/climber/commands/AutoClimb.java

Co-authored-by: orel-david <42713127+orel-david@users.noreply.github.com>

natymeitav committed:
Format Intake Ports.

Co-authored-by: Barel Shlidor <42340005+barel13@users.noreply.github.com>

natymeitav committed:
Rename IS_INVERTED to IS_MOTOR_INVERTED.

Co-authored-by: Dan Katzuv <31829093+katzuv@users.noreply.github.com>

natymeitav committed:
Rename SOLENOID_RIGHT & SOLENOID_LEFT to RIGHT_SOLENOID & LEFT_SOLENOID.

Co-authored-by: Dan Katzuv <31829093+katzuv@users.noreply.github.com>

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,251 @@
+            controls[i][1] = polar[1];
+        }
+
+        // feeds the corresponding control to each wheel
+        for (int k = 0; k < 4; k++) {
+            swerveModules[k].setSpeed(controls[k][0]);
+            swerveModules[k].setAngle(controls[k][1]);
+        }
```
Why not use this in the above for loop? (then you can also remove the controls 2d array)

adamv17 replied:
I did this on purpose to make the code more understandable and "highlight" the controls and if needed it will be easy to check their values. If you think this is unnecessary I can change it.

natymeitav committed:
Add requirements to StartFunnel.

natymeitav committed:
Rename direction to isMovingUp.

natymeitav committed:
Rename setVelocity to setPower.

natymeitav committed:
Shorten startIntake's code.

Co-authored-by: Barel Shlidor <42340005+barel13@users.noreply.github.com>

Se-gol committed:
More fixes to climber.

natymeitav committed:
Make Intake control one solenoid.

Se-gol committed:
Remove unnecessary variable from PistonMode enum.

adamv17 committed:
Add method

adamv17 committed:
Utilize method

adamv17 committed:
Replace command class with inline command

orel-david committed:
Fix typo at the name

adamv17 committed:
Optimize imports

orel-david committed:
Improve the switch between the subsystems

orel-david committed:
Document the command

orel-david committed:
Document methods

barel13 commented on src/main/java/frc/robot/subsystems/drivetrain/SwerveDrive.java:
```diff
@@ -0,0 +1,260 @@
+        double[][] rotationMat = {{Math.cos(robotAngle), -Math.sin(robotAngle)},
+                {Math.sin(robotAngle), Math.cos(robotAngle)}};
+        double[] speeds = Utils.matrixVectorMult(rotationMat, new double[]{forward, strafe});
+        System.out.println("forward" + forward);
+        System.out.println("strafe" + strafe);
```
do we need this print? if we want to see the values we should use the FireLog, am I wrong?

adamv17 replied:
you are correct this was a temporary print during testing

Se-gol committed:
Update src/main/java/frc/robot/subsystems/climber/Climber.java

Co-authored-by: Barel Shlidor <42340005+barel13@users.noreply.github.com>

Se-gol committed:
Final fixes to Climber.

Se-gol committed:
Merge remote-tracking branch 'origin/climber' into climber

orel-david committed:
Merge remote-tracking branch 'origin/climber' into integration

orel-david committed:
Merge shooter

orel-david committed:
Add instance of the pto to the robot container and fix merging errors

orel-david committed:
Move the motors configuration to the PTO

orel-david committed:
Rename the motors

orel-david committed:
Use the Pto to get the motors

orel-david committed:
Reformat

barel13 committed:
Use 2 proximity sensors in the DeadbandProximity

orel-david committed:
Fix

orel-david committed:
Fix

orel-david committed:
Add color sensor

orel-david committed:
Add port and change command

orel-david committed:
Invert motor

adamv17 committed:
Merge remote-tracking branch 'origin/intake' into pipeline-integration

adamv17 committed:
Fix conflicts

adamv17 committed:
Merge remote-tracking branch 'origin/funnel' into pipeline-integration

adamv17 committed:
Add spark max gradle

adamv17 committed:
Merge remote-tracking branch 'origin/conveyor' into pipeline-integration

adamv17 committed:
Format

adamv17 committed:
Add command group

Commit:
Fix errors

Commit:
Add VisionModule and fix swerve

Commit:
Change motor to Falcon

Commit:
Add code to turn to target

Commit:
Change navx getter to yaw

Commit:
Update Orel

Commit:
Create VisionModule class

Commit:
Add the upper motor to the shooter

Commit:
Use the upper motor in shoot command

Commit:
Make the velocity constant public & static

Commit:
Add navx to the Robot

Commit:
Add up port to Ports file

Commit:
General fixes

Commit:
Create UtilityFunctions class

Commit:
Remove gearbox from Climber

Commit:
Fix climber modes

Commit:
General fixes

Commit:
Add drum to climber

Commit:
Change upMotor velocity

Commit:
Add TurnOffLeds command

adamv17 committed:
Merge remote-tracking branch 'origin/pipeline-integration' into swerve

adamv17 committed:
Fix conflicts

adamv17 committed:
Add MoveToPosition command
